### PR TITLE
Add distinct And distinctOn Functions To QueryBuilder

### LINKED ISSUE #654: Add Distinct Functions For QueryBuilder
- `distinctOn #priority` would map to SQL `DISTINCT ON priority`

### PR DESCRIPTION
**Pull request by s0kil**
Hacked this one together thanks to #443, #632, and #434

Fixes #654

### CODE REVIEW

**Comment by s0kil** [comment]
I'm not sure if DISTINCT ON with multiple values in valid, `SELECT DISTINCT ON (body, text)`?

File: IHP/QueryBuilder.hs [hunk 1]
```diff
@@ -268,11 +276,13 @@ toSQL :: forall table. (KnownSymbol table) => QueryBuilder table -> (ByteString,
...
 {-# INLINE toSQL #-}
 
-toSQL' sqlQuery@SQLQuery { selectFrom, orderByClause, limitClause, offsetClause } =
+toSQL' sqlQuery@SQLQuery { selectFrom, distinctClause, distinctOnClause, orderByClause, limitClause, offsetClause } =
         (DeepSeq.force theQuery, theParams)
     where
         !theQuery =
-            "SELECT " <> selectors <> " FROM "
+            "SELECT "
+            <> distinctClause' <> distinctOnClause' <> " "
```
Comment (s0kil) [inline IHP/QueryBuilder.hs:284]: There is probably a way to merge these two into one, but that is beyond me at the moment.

File: IHP/QueryBuilder.hs [hunk 2]
```diff
@@ -512,4 +524,31 @@ instance (model ~ GetModelById (Id' table), value ~ Id' table, HasField "id" mod
...
+-- Use 'distinct' to remove all duplicate rows from the result
+--
+-- __Example:__ Fetch distinct books
+--
+-- > query @Book
+-- >     |> distinct
+-- >     |> fetch
+-- > -- SELECT DISTINCT * FROM books
+distinct :: QueryBuilder table -> QueryBuilder table
+distinct = DistinctQueryBuilder
```
Comment (mpscholten) [inline IHP/QueryBuilder.hs:540]:
```suggestion
distinct = DistinctQueryBuilder
{-# INLINE distinct #-}
```

**Review by mpscholten — COMMENTED** [review]
Looks good 👍 

Can we add `Fetching distinct records` to https://ihp.digitallyinduced.com/Guide/database.html#retrieving-records ? Otherwise people might not know that we have these functions :)

**Review by fegu — COMMENTED** [review]
Looks good. Nice addition.

File: IHP/QueryBuilder.hs (hunk 2 above)
Comment (fegu) [reply IHP/QueryBuilder.hs:540]: There's a lot of explicit inlining in the IHP codebase. How much effect does this actually have in this day & age?

**Comment by s0kil** [comment]
I'm added the changes requested.

File: IHP/QueryBuilder.hs (hunk 2 above)
Comment (mpscholten) [reply IHP/QueryBuilder.hs:540]: We'd need to benchmark this at some point (but that's low prio). In general I guess it helps to specialize type heavy code (like query builder) and avoids passing too many class dicts around.

**Review by mpscholten — APPROVED** [review]
Perfect, thanks :)

**Pull request merged**